File: 0024-swap-nodes-in-pairs/0024-swap-nodes-in-pairs.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var swapPairs = function(head) {
  if (!head) return null;

  const result = [];
  let temp;
  let currentNode = head;

  while (currentNode) {
    if (temp) {
      result.push(currentNode);
      result.push(temp);
      temp = null;
    } else {
      temp = currentNode;
    }
    currentNode = currentNode.next;
  }

  if (temp) {
    result.push(temp);
  }

  for (let i = 0; i < result.length - 1; i++) {
    result[i].next = result[i + 1];
    result[i + 1].next = null;
  }

  return result[0];
};

File: 0024-swap-nodes-in-pairs/0024-swap-nodes-in-pairs.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dummy_list = ListNode(None, head)
        first = dummy_list
        second = first

        while second:
            if first.next is None or first.next.next is None:
                return dummy_list.next
            
            second = first.next.next
            temp = second.next
            second.next = first.next
            first.next.next = temp
            first.next = second
            first = second = first.next.next

        return dummy_list.next